@xen-orchestra/xapi/vdi.mjs

```diff
@@ -95,96 +96,114 @@ class Vdi {
-            `Exporting content of VDI ${vdiName} using NBD ${changedBlocks !== undefined ? ' and CBT' : ''}`
-          )
-          exportStream = stream
-          stream = await createNbdVhdStream(nbdClient, exportStream, {
-            changedBlocks,
-            vdiInfos: { size, uuid, parentUuid: baseParentUuid },
-          })
-          stream.on('progress', progress => this.call('task.set_progress', taskRef, progress))
-          finished(stream, () => {
-            nbdClient.disconnect()
-            exportStream?.destroy() // ensure the source stream is really closed
-
-            this.task_destroy(taskRef).catch(warn)
-          })
+    // really connect to NBD server
+    if (preferNbd) {
+      nbdClient = await this._getNbdClient(ref, { nbdConcurrency })
+      // disconnect on failure or when transfer is finished
+      $defer.onFailure(() => nbdClient?.disconnect())
+    }
+
+    // create a xapi export stream only if we won't make a delta from CBT
+    // a CBT export can only work if we have a NBD client and changed blocks
+    if (changedBlocks === undefined || nbdClient === undefined) {
+      stream = exportStream = (
+        await this.getResource(cancelToken, '/export_raw_vdi/', {
+          query,
+          task: await this.task_create(`Exporting content of VDI ${vdiName}`),
+        })
+      ).body
+      exportStream.on('error', err => {
+        // this error code is directly caused by exportStream.destroy
+        if (err.code !== 'UND_ERR_ABORTED') {
+          throw err
         }
-      }
-      return stream
-    } catch (error) {
-      // augment the error with as much relevant info as possible
-      const [poolMaster, vdi] = await Promise.all([
-        this.getRecord('host', this.pool.master),
-        this.getRecord('VDI', ref),
-      ])
-      error.pool_master = poolMaster
-      error.SR = await this.getRecord('SR', vdi.SR)
-      error.VDI = vdi
-      error.nbdClient = nbdClient
-      nbdClient?.disconnect()
+      })
+
+      $defer.onFailure(() => {
+        exportStream.destroy()
+      })
+    }
+
+    // we have a NBD client : use it to transfer download data
+    // use either the changed block or exportStream to compute the header/footer/BAT
+    if (nbdClient !== undefined) {
+      taskRef = await this.task_create(
+        `Exporting content of VDI ${vdiName} using NBD ${changedBlocks !== undefined ? ' and CBT' : ''}`
+      )
+      $defer.onFailure(() => this.task_destroy(taskRef).catch(warn))
+      // stream is now using CBT, exportStream still keep a ref to the XAPI export stream
+      stream = await createNbdVhdStream(nbdClient, exportStream, {
+        changedBlocks,
+        vdiInfos: { size, uuid, parentUuid: baseParentUuid },
+        onProgress: async progress => {
+          this.call('task.set_progress', taskRef, progress).catch(warn)
+        },
+      })
+      // we don't need the export stream anymore, block data will be read through NBD
       exportStream?.destroy()
```

Comment:
What happens if the next `assert` fails, won't `exportStream` be destroyed twice (once here and once in the `defer.onFailure()`)?

Reply:
it's not really a problem since 
> Once destroy() has been called any further calls will be a no-op and no further errors except from _destroy() may be emitted as 'error'.